obp/ope/estimators_embed.py

```diff
@@ -0,0 +1,666 @@
+
+        action_dist: array-like, shape (n_rounds, n_actions, len_list)
+            Action choice probabilities of the evaluation policy (can be deterministic), i.e., :math:`\\pi_e(a_i|x_i)`.
+
+        position: array-like, shape (n_rounds,), default=None
+            Indices to differentiate positions in a recommendation interface where the actions are presented.
+            If None, the effect of position on the reward will be ignored.
+            (If only a single action is chosen for each data, you can just ignore this argument.)
+
+        p_e_a: array-like, shape (n_actions, n_cat_per_dim, n_cat_dim), default=None
+            Conditional distribution of action embeddings given each action.
+            This distribution is available only when we use synthetic bandit data, i.e.,
+            `obp.dataset.SyntheticBanditDatasetWithActionEmbeds`.
+            See the output of the `obtain_batch_bandit_feedback` argument of this class.
+            If `p_e_a` is given, MIPW uses the true marginal importance weights based on this distribution.
+            The performance of MIPW with the true weights is useful in synthetic experiments of research papers.
+
+        with_dev: bool, default=False.
+            Whether to output a deviation bound with the estimated sample-wise rewards.
+
+        Returns
+        ----------
+        estimated_rewards: array-like, shape (n_rounds,)
+            Estimated rewards for each observation.
+
+        """
+        n = reward.shape[0]
+        if p_e_a is not None:
+            p_e_pi_b = np.ones(n)
+            p_e_pi_e = np.ones(n)
+            for d in np.arange(p_e_a.shape[-1]):
+                p_e_pi_b_d = pi_b[np.arange(n), :, position] @ p_e_a[:, :, d]
+                p_e_pi_b *= p_e_pi_b_d[np.arange(n), action_embed[:, d]]
+                p_e_pi_e_d = action_dist[np.arange(n), :, position] @ p_e_a[:, :, d]
+                p_e_pi_e *= p_e_pi_e_d[np.arange(n), action_embed[:, d]]
+            w_x_e = p_e_pi_e / p_e_pi_b
+            self.max_w_x_e = w_x_e.max()
+
+        else:
+            w_x_e = self._estimate_w_x_e(
+                context=context,
+                action=action,
+                action_embed=action_embed,
+                pi_e=action_dist[np.arange(n), :, position],
+                pi_b=pi_b[np.arange(n), :, position],
+            )
+            self.max_w_x_e = w_x_e.max()
+
+            if with_dev:
+                r_hat = reward * w_x_e
+                cnf = np.sqrt(np.var(r_hat) / (n - 1))
+                cnf *= stats.t.ppf(1.0 - (self.delta / 2), n - 1)
+
+                return r_hat.mean(), cnf
+
+        return reward * w_x_e
+
+    def _estimate_w_x_e(
+        self,
+        context: np.ndarray,
+        action: np.ndarray,
+        action_embed: np.ndarray,
+        pi_b: np.ndarray,
+        pi_e: np.ndarray,
+    ) -> np.ndarray:
+        """Estimate the marginal importance weights."""
```

Comment:
Thanks for pointing this out. First, this is not related to B-OPE. I implemented a procedure to estimate the marginal importance weight `w(x,e)` described in Section 3.3 of https://arxiv.org/pdf/2202.06317.pdf

In order to estimate `w(x,e)`, we first estimate `\pi(a|x,e)`, which is `p_a_e` in the code. Then, we take the expectation of `w(x,a)` with respect to `\hat{\pi}(a|x,e)`. (Eq.(2) of the paper).

Reply:
> 1. Linear Logistic model cannot capture interactions between contexts and actions

Thank you! That might be true. I modified the implementation a bit, and now users can choose any classifier such as RandomForest as `p_a_e_model`. (Actually, I also renamed `p_a_e_model` to `p_a_e_estimator`).


> 2. B-OPE should use Binary classification, but multi-class classification seems to be..

This is not B-OPE, but rather estimates `\pi(a|x,e)` via multi-class classification.

> 3. I cannot understand the last summation ( w_x_e = (w_x_a * p_a_e).sum(1) )

This takes the expectation of w(x,a) with respect to \hat{\pi}(a|x,e). (Eq.(2) of the paper).

Reply:
OK, I understand the implementation and modification!!

@usaito 

[imo]

It would be better to add some (performance) tests when `p_e_a` is None and `embedding_selection_method` is not None in order to check the validity of the implementations.

Reply:
@fullflu Thanks for checking! Do you mean I should modify https://github.com/st-tech/zr-obp/blob/feature/marginal-ips/tests/ope/test_offline_estimation_performance.py ?

Reply:
> Do you mean I should modify https://github.com/st-tech/zr-obp/blob/feature/marginal-ips/tests/ope/test_offline_estimation_performance.py ?

Yes!

Or, it would be sufficient to add estimators where `embedding_selection_method` is not None in: https://github.com/st-tech/zr-obp/blob/feature/marginal-ips/tests/ope/test_ipw_estimators_embed.py

Reply:
@fullflu Sure! will do from now